### ISSUE TITLE
"@azure/functions-core" can't be resolved by bundlers

### ISSUE THREAD

**Issue by pcj** [Azure/azure-functions-nodejs-library#201]
When using a bundler such as `esbuild`, bundling `@azure/functions` causes this error:

```js
✘ [ERROR] Could not resolve "@azure/functions-core"

    node_modules/@azure/functions/dist/azure-functions.js:1657:25:
      1657 │ module.exports = require("@azure/functions-core");
```

> NOTE: bundling with `ncc` has same problem but `esbuild` emits a better error message

There is no published npm module `@azure/functions-core`.

I see that there is a path mapping in `tsconfig.json`, perhaps it's a shim for old or MS internal code?

In any case, this breaks bundlers and is very annoying.  Can we please get this removed from the tsconfig and just use the correct internal import reference?

**Comment by pcj** [comment]
```json
{
  "scripts": {
      "compile": "tsc",
      "bundle": "esbuild dist/src/azure/main.js --bundle --platform=node --external:@azure/functions-core --outfile=dist/index.js",
  },
  "main": "dist/index.js"
}
```

This is what I'm using, hope it's correct.

**Comment by ejizba** [comment]
Hi @pcj, "@azure/functions-core" is a built-in module that ships directly in Azure, not through npm. I understand this breaks bundlers, but it's an important piece that links your app code to the underlying Azure Functions runtime. It's analogous to how "path" is a built-in module that ships directly in Node.js. In both cases, the Azure Functions runtime and the Node.js runtime are providing functionality unique to their environment through these modules.

Most of our customers actually don't need to bundle their app. We highly recommend you [run from a package file](https://learn.microsoft.com/en-us/azure/azure-functions/run-functions-from-deployment-package), which provides the performance benefits of bundling and is a lot easier to configure.

If you still need to use a bundler, then yes the recommendation is to list "@azure/functions-core" as an external package so that your bundler will ignore it.

**Comment by pcj** [comment]
@ejizba Thank you for a prompt response.  

Personally I think "running from a package file" is somewhat orthogonal to this issue, because the customer is still responsible for creating the zip file, which in the naive case just zipping up everything including the `node_modules/` directory.

In my case, that would be difference of having a deployment zip file around `725M` rather than `5M`.  So, deployments would be much slower.

Maybe I missed something about kudu or some other thing regarding zip deployments, because from where I am standing there is no built-in performance benefit.

I personally do recommend a bundler for this reason.

But I'll go ahead and close the issue.  Hopefully it helps someone else in the future.

**Comment by pcj** [comment]
Also, your comment:

> "@azure/functions-core" is a built-in module that ships directly in Azure, not through npm.

Is very helpful, I didn't find this documented elsewhere.

**Comment by ejizba** [comment]
You're correct it would not help with deployment performance at all. I was talking about performance of running your app, specifically every time an instance is started which loads your app code. Instances can be recycled often so the startup performance is definitely important.

Another option for speeding up deployments is to run `npm prune --production` right before you deploy. I find that dev dependencies often account for a large amount of the "node_modules" folder size. But if you already got your bundler working and are happy with that, I see no reason to stop using it.

> Is very helpful, I didn't find this documented elsewhere.

This was a relatively recent thing starting with the new programming model that just GA-ed in September. We knew the core module wouldn't work with bundlers, but you're the first person to bring it up. I expect you won't be the last, so I can re-open this issue to track doc improvements.

**Comment by edgehero** [comment]
this is a major bug on @azure/functions what makes it not possible to use on monorepo's or any bigger projects. when will this be fixed?

**Comment by ejizba** [comment]
Hi @edgehero, please create a new issue with detailed repro steps including how you set up your monorepo, what tools you're using, and what errors you're getting: https://github.com/Azure/azure-functions-nodejs-library/issues/new

This issue is only open to track documentation updates, specifically to make it more obvious that folks should list `@azure/functions-core` as an external dependency in their bundler. If that recommendation doesn't work in a monorepo, I'd be happy to follow up once we get those details

**Comment by mildronize** [comment]
_Moved to https://github.com/Azure/azure-functions-nodejs-library/issues/260_

**Comment by edgehero** [comment]
_Moved to https://github.com/Azure/azure-functions-nodejs-library/issues/260_

**Comment by ejizba** [comment]
@mildronize @edgehero I've moved your comments to a new issue: https://github.com/Azure/azure-functions-nodejs-library/issues/260

The correct action here is to list `@azure/functions-core` as an external dependency and this is a common practice that should work with any repo configuration. Your problems are likely specific to your own setup and do not reflect a bug on Azure Functions. I'm happy to help you troubleshoot together in a separate issue

**Comment by igorpupkinable** [comment]
> Also, your comment:
> 
> > "@azure/functions-core" is a built-in module that ships directly in Azure, not through npm.
> 
> Is very helpful, I didn't find this documented elsewhere.

```ts
/**
 * Intercepts the default "require" method so that we can provide our own "built-in" module
 * This module is essentially the publicly accessible API for our worker
 * This module is available to users only at runtime, not as an installable npm package
 */
```

https://github.com/Azure/azure-functions-nodejs-worker/blob/v3.x/src/setupCoreModule.ts#L36

**Comment by igorpupkinable** [comment]
@ejizba I was expecting `@azure/functions-core` to be available during runtime and configured externals as follows:
```ts
  externals: [
    '@azure/functions-core',
    '@azure/functions-extensions-base',
    'cookie',
  ],
```

However `@azure/functions-core` is not available during runtime.
```bash
Azure Functions Core Tools
Core Tools Version:       4.9.0+29aeab590e6c229a1a6631f6653df70f150f9f0e (64-bit)
Function Runtime Version: 4.1047.100.26071

Worker was unable to load entry point "dist/index.mjs": Cannot find package '@azure/functions-core' imported from /[REDACTED]/dist/index.mjs
Error [ERR_MODULE_NOT_FOUND]: Worker was unable to load entry point "dist/index.mjs": Cannot find package '@azure/functions-core' imported from /[REDACTED]/dist/index.mjs
     at Object.getPackageJSONURL (node:internal/modules/package_json_reader:314:9)
     at packageResolve (node:internal/modules/esm/resolve:768:81)
     at moduleResolve (node:internal/modules/esm/resolve:855:18)
     at defaultResolve (node:internal/modules/esm/resolve:985:11)
     at #cachedDefaultResolve (node:internal/modules/esm/loader:731:20)
     at ModuleLoader.resolve (node:internal/modules/esm/loader:708:38)
     at ModuleLoader.getModuleJobForImport (node:internal/modules/esm/loader:310:38)
     at ModuleJob._link (node:internal/modules/esm/module_job:182:49)
     at process.processTicksAndRejections (node:internal/process/task_queues:103:5)
```

How do I make it available locally with `yarn run func start dist/index.mjs`?
